File: widgetssdk/src/main/java/com/glia/widgets/helper/TimeCounter.java
package com.glia.widgets.helper;

import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.TimeUnit;

public class TimeCounter {

    private final static String TAG = "TimeCounter";
    private TimerTask timerTask;
    private Timer timer;
    private final List<FormattedTimerStatusListener> formattedListeners = new ArrayList<>();
    private final List<RawTimerStatusListener> rawListeners = new ArrayList<>();

    public void startNew(int timerDelayMs, int timerIntervalMs) {
        Logger.d(TAG, "startNew! timerDelay: " + Integer.valueOf(timerDelayMs).toString() +
                ", timerInterval: " + Integer.valueOf(timerIntervalMs).toString());
        createNewTimerTask(timerDelayMs, timerIntervalMs);
        createNewTimer();
        timer.schedule(timerTask, timerDelayMs, timerIntervalMs);
    }

    public void stop() {
        Logger.d(TAG, "stop");
        if (timerTask != null) {
            timerTask.cancel();
            timerTask = null;
        }
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
    }

    public void clear() {
        Logger.d(TAG, "clear");
        stop();
        formattedListeners.clear();
        rawListeners.clear();
    }

    public void addFormattedValueListener(FormattedTimerStatusListener listener) {
        Logger.d(TAG, "addFormattedListener");
        formattedListeners.add(listener);
    }

    public void removeFormattedValueListener(FormattedTimerStatusListener listener) {
        Logger.d(TAG, "removeFormattedListener");
        formattedListeners.remove(listener);
    }

    public void addRawValueListener(RawTimerStatusListener listener) {
        Logger.d(TAG, "addRawListener");
        rawListeners.add(listener);
    }

    private void createNewTimer() {
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
        timer = new Timer();
    }

    private void createNewTimerTask(int timerDelayMs, int timerIntervalMs) {
        if (timerTask != null) {
            timerTask.cancel();
            timerTask = null;
        }
        timerTask = new TimerTask() {
            int value = timerDelayMs;

            @Override
            public void run() {
                Logger.d(TAG, "Timer value: " + value);
                for (RawTimerStatusListener listener : rawListeners) {
                    listener.onNewRawTimerValue(value);
                }

                if (!formattedListeners.isEmpty()) {
                    String time = Utils.toMmSs(Long.valueOf(TimeUnit.MILLISECONDS.toSeconds(value)).intValue());
                    Logger.d(TAG, "Formatted timer: " + time);
                    for (FormattedTimerStatusListener listener : formattedListeners) {
                        listener.onNewFormattedTimerValue(time);
                    }
                }
                value = value + timerIntervalMs;
            }

            @Override
            public boolean cancel() {
                for (RawTimerStatusListener listener : rawListeners) {
                    listener.onRawTimerCancelled();
                }
                for (FormattedTimerStatusListener listener : formattedListeners) {
                    listener.onFormattedTimerCancelled();
                }
                Logger.d(TAG, "cancel");
                return super.cancel();
            }
        };
    }

    public boolean isRunning() {
        return timerTask != null;
    }

    public interface FormattedTimerStatusListener {
        void onNewFormattedTimerValue(String formatedValue);

        void onFormattedTimerCancelled();
    }

    public interface RawTimerStatusListener {
        void onNewRawTimerValue(int timerValue);

        void onRawTimerCancelled();
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/helper/TimeCounter.kt
package com.glia.widgets.helper

import java.util.Timer
import java.util.TimerTask

internal class TimeCounter {
    private val formattedListeners: MutableList<FormattedTimerStatusListener> = ArrayList()
    private val rawListeners: MutableList<RawTimerStatusListener> = ArrayList()
    private var timerTask: TimerTask? = null
    private var timer: Timer? = null
    fun startNew(timerDelayMs: Int, timerIntervalMs: Int) {
        Logger.d(TAG, "startNew! timerDelay: " + Integer.valueOf(timerDelayMs).toString() +
                    ", timerInterval: " + Integer.valueOf(timerIntervalMs).toString())
        createNewTimerTask(timerDelayMs, timerIntervalMs)
        createNewTimer()
        timer?.schedule(timerTask, timerDelayMs.toLong(), timerIntervalMs.toLong())
    }

    fun stop() {
        Logger.d(TAG, "stop")
        timer?.cancel()
        timer = null

        timerTask?.cancel()
        timerTask = null
    }

    fun clear() {
        Logger.d(TAG, "clear")
        stop()
        formattedListeners.clear()
        rawListeners.clear()
    }

    fun addFormattedValueListener(listener: FormattedTimerStatusListener) {
        Logger.d(TAG, "addFormattedListener")
        formattedListeners.add(listener)
    }

    fun removeFormattedValueListener(listener: FormattedTimerStatusListener) {
        Logger.d(TAG, "removeFormattedListener")
        formattedListeners.remove(listener)
    }

    fun addRawValueListener(listener: RawTimerStatusListener) {
        Logger.d(TAG, "addRawListener")
        rawListeners.add(listener)
    }

    private fun createNewTimer() {
        timer?.cancel()
        timer = Timer()
    }

    private fun createNewTimerTask(timerDelayMs: Int, timerIntervalMs: Int) {
        timerTask?.cancel()
        timerTask = object : TimerTask() {
            var value = timerDelayMs

            @Synchronized
            override fun run() {
                Logger.d(TAG, "Timer value: $value")
                for (listener in rawListeners) {
                    listener.onNewRawTimerValue(value)
                }

                if (formattedListeners.isNotEmpty()) {
                    val time = formatElapsedTime(value.toLong())
                    Logger.d(TAG, "Formatted timer: $time")
                    for (listener in formattedListeners) {
                        listener.onNewFormattedTimerValue(time)
                    }
                }
                value += timerIntervalMs
            }

            @Synchronized
            override fun cancel(): Boolean {
                for (listener in rawListeners) {
                    listener.onRawTimerCancelled()
                }
                for (listener in formattedListeners) {
                    listener.onFormattedTimerCancelled()
                }
                Logger.d(TAG, "cancel")
                return super.cancel()
            }
        }
    }

    val isRunning: Boolean
        get() = timerTask != null

    interface FormattedTimerStatusListener {
        fun onNewFormattedTimerValue(formattedValue: String)
        fun onFormattedTimerCancelled()
    }

    interface RawTimerStatusListener {
        fun onNewRawTimerValue(timerValue: Int)
        fun onRawTimerCancelled()
    }
}
